### ISSUE TITLE
Some wrongs in the codes and problems in reproduction

### ISSUE THREAD

**Issue by Guyujie666** [MediaBrain-SJTU/FedLESAM#1]
Thanks for your wonderful implementations of such a lot of FL methods! It is really helpful! But I find there are some wrongs in line 48 in file 'LESAM.py', since the 'g_update' is a vector of length 11181642 while the 'idx' varies from 0 to 61, which is the index of layers of resnet. I guess you meant to reshape the 'g_update' vector to the shape of each layer in the resnet model but you might forget. The same wrong is also in 'LESAM-S.py' and 'LESAM-D.py'. What's more, I have tried to reproduce your results in the paper but FedLESAM seems to perform worse than FedSAM (76.40% for FedLESAM and 77.51% for FedSAM) and the loss landscape of FedLESAM seems sharper than FedSAM by drawing with the python package 'loss_landscapes' ( the transparent loss landscape in the below is FedLESAM's and the other is FedSAM's). I use the hyperparameters recommended in your paper as follows: 'python train.py --non-iid --dataset CIFAR10 --model ResNet18 --split-rule Dirichlet --split-coef 0.1 --active-ratio 0.1 --total-client 100 --batchsize 50 --rho 0.01 --method FedLESAM --lr-decay 0.998 --local-epochs 5 --comm-rounds 800 --local-learning-rate 0.1' and only change method 'FedLESAM' to 'FedSAM' for FedSAM. Can you help me solve the problems above? Thank you very much.
![lesam](https://github.com/user-attachments/assets/2e74f140-698c-479c-a55f-4cee487256be)

**Comment by knight-fzq** [comment]
You may try larger rho for FedLESAM. For example rho=0.5.

**Comment by knight-fzq** [comment]
You can contact me at zqfan_knight@sjtu.edu.cn. I will be able to see your questions more promptly this way.

**Comment by harrylee999** [comment]
@Guyujie666 heybro，how is this problem going? i am facing the same problem,  could u share your solutions  if you have? Thanks!

**Comment by Guyujie666** [comment]
Hi, bro! Sorry for that I have not check my QQ email for a long time. I have gave up for a month because no good results exist with my great efforts. Have you solved the problem?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "MediaBrain-SJTU/FedLESAM"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年1月13日(星期一) 下午5:07
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [MediaBrain-SJTU/FedLESAM] Some wrongs in the codes and problems in reproduction (Issue #1)





 
@Guyujie666 heybro，how is this problem going? i am facing the same problem,  could u share your solutions  if you have? Thanks!

**Comment by harrylee999** [comment]
> Hi, bro! Sorry for that I have not check my QQ email for a long time. I have gave up for a month because no good results exist with my great efforts. Have you solved the problem?
> […](#)

Same...，I can not reproduce the result as good as the original paper. I find  FedSAM is higher 0%~1% Acc than FedLESAM in most settings and FedSMOO is better  1%~2% Acc than FedLESAM-D/S in most cases.

**Comment by Guyujie666** [comment]
same.. LESAM never exceeds FedSAM in my reproduction in various settings... I even run the setting that only one step update at each client and full participation, which can confirm the validation of the estimated global perturbation in LESAM to some degree if it has improvement. But the test accuracy is worse than FedSAM :(

**Comment by knight-fzq** [comment]
Hi all, it seems there is a bug in Line48 e_w=-g_update[idx] * scale.to(p). Here g_update is a tensor instead of a dict as expected, due to some personal errors. You may use "gradient tensor to gradient dict" function transforming it to desired format.
